tests/test_rewrite_rules/test_rewrite_rules.py

```diff
@@ -32,141 +29,134 @@ class TestEntry:
-                return test_it(entry, default_header)
-            else:
-                r1 = test_it(entry, entry.accept_header)
-                return test_it(entry, entry.accept_header + "," + default_header) and r1
+    return {
+        "meta_entries": meta_entries,
+        "vocab_entries": vocab_entries,
+        "meta_model_entries": meta_model_entries,
+        "meta_vocab_entries": meta_vocab_entries,
+    }
 
-        self.assertTrue(all([ev_al(entry) for entry in entries]))
 
-    def test_type_meta(self):
-        self.rule_test(self.tests.meta_entries)
+@pytest.fixture
+def fail_on_error():
+    return os.environ.get("FAIL_ON_ERROR", "False") == "True"
 
-    def test_type_entry(self):
-        self.rule_test(self.tests.vocab_entries)
 
-    def test_meta_meta(self):
-        self.rule_test(self.tests.meta_model_entries)
+@pytest.fixture(scope="class")
+def results() -> Set[Tuple[str, str, str]]:
+    return set()  # from, to, format
 
-    def test_meta_entry(self):
-        self.rule_test(self.tests.meta_vocab_entries)
 
+def record_results(results, from_url: str, accept_header, to_url: str) -> None:
+    results.add((from_url, to_url, accept_header.split(",")[0]))
 
-if __name__ == "__main__":
-    RewriteRuleTestCase.SERVER = os.environ.get("SERVER", RewriteRuleTestCase.SERVER)
-    unittest.main()
+
+fixture_lists = generate_fixture_lists()
+
+
+@pytest.mark.network
+@pytest.mark.skipif(SKIP_REWRITE_RULES, reason=SKIP_REWRITE_RULES_REASON)
+@pytest.mark.parametrize(
+    "entries",
+    [
+        pytest.param(fixture_lists["meta_entries"], id="meta_entries"),
+        pytest.param(fixture_lists["vocab_entries"], id="vocab_entries"),
+        pytest.param(fixture_lists["meta_model_entries"], id="meta_model_entries"),
+        pytest.param(fixture_lists["meta_vocab_entries"], id="meta_vocab_entries"),
+    ],
+)
+def test_rewrite_rules(entries: List[TestEntry], results, fail_on_error) -> None:
+    def test_it(e: TestEntry, accept_header: str) -> bool:
+        expected = github_io + e.expected_url
+        resp = requests.head(e.input_url, headers={"accept": accept_header}, verify=False)
+
+        # w3id.org uses a 301 to go from http: to https:
+        if resp.status_code == 301 and "location" in resp.headers:
+            resp = requests.head(
+                resp.headers["location"],
+                headers={"accept": accept_header},
+                verify=False,
+            )
+        actual = (
+            resp.headers["location"]
+            if resp.status_code == 302 and "location" in resp.headers
+            else f"Error: {resp.status_code}"
+        )
+        if fail_on_error:
+            assert expected == actual, f"redirect for: {resp.url}"
+            record_results(results, e.input_url, accept_header, actual)
+            return True
+        elif expected != actual:
+            print(f"{e.input_url} ({accept_header}):\n expected {expected} - got {actual}")
+            return False
```

Comment:
antipattern - test result should be meaningful and uniform. A failing test only fails, and a passing test only passes. printing rather than raising (and generally making tests lenient and not emit expected output in CI) has been the cause of a ton of hard to diagnose problems.

tests should never print - pytest shouldn't emit to stdout, and if we are running tests such that they do output to stdout, that's a sign we're running things outside of pytest and thus not in a consistent environment. the printed information can be gotten by looking at the pytest summary, and one can run individual test modules and individual tests themselves with pytest (i.e. we should never run tests by just running them as python modules).

Reply:
This was mostly a whitespace change when I pulled the test methods out of the unittest class. I also changed FAIL_ON_ERROR to the fail_on_error() fixture for more pyunit style. Git diffs make it seem like a bigger change.

That said, I should have changed the prints to logging per the issue.
I did consider removing this whole section (and still can if that's preferred).